### PR TITLE
Quick Edit: add Template field

### PR DESCRIPTION
**Pull request by gigitux**
## What?

This PR is based on top of https://github.com/WordPress/gutenberg/pull/66531
This PR is part of https://github.com/WordPress/gutenberg/issues/64519.

## Why?

This PR adds the `Template` field.

## Testing Instructions

1. Enable the DataViews specific experiments under Gutenberg > Experiments.
2. Make sure you have a block themed theme enabled.
3. Go to Appearance > Editor and click Pages.
4. Switch the page view to the table view.
5. Select a page and select the Details icon.
6. A Template field should now be displayed.
7. Click on the template field and ensure that a dropdown is visible.
8. Click on swap template.
9. Ensure that a modal is visible and all the templates are rendered correctly.
10. Click on a template.
11. Save.
12. Ensure that the page has as template the selected one.
13. Click on the template field and ensure that a dropdown is visible.
14. Click on `Use default template`.
15. Ensure that the default template is set.

## Screenshots or screencast 

https://github.com/user-attachments/assets/5b278d3f-2f54-4465-a91a-d365a5d1b73c

### `@wordpress/block-editor` dependency

To implement the modal previewing all the templates is necessary use the `__experimentalBlockPatternsList` from the `@wordpress/block-editor` package. I think that it is acceptable.

### ⚠️ `Uncaught TypeError: Cannot destructure property 'avatarURL' of '(0 , _wordpress_data__WEBPACK_IMPORTED_MODULE_3__.useSelect)(...)' as it is undefined` error

In some cases, the templates in the modal aren't rendered correctly due to this JavaScript error:

```
Uncaught TypeError: Cannot destructure property 'avatarURL' of '(0 , _wordpress_data__WEBPACK_IMPORTED_MODULE_3__.useSelect)(...)' as it is undefined
```

https://github.com/user-attachments/assets/7950309e-31ce-4c2f-a173-cf9825fd4839

https://github.com/WordPress/gutenberg/blob/61f965533542941fd9ffe525933fa61c52464f00/packages/block-library/src/avatar/hooks.js#L19-L26

This happens when the setting object isn't initialized. To fix this issue, I pushed 5a623fbf0825298547b80f782e420610e6d61bab. Still, I'm not sure that this is the best solution 🤔.

It is a similar issue that I tried to tackle with #66459.

cc @oandregal @youknowriad and @louwie17

### CODE REVIEW

**Comment by gigitux** [comment]
I rebased this PR after the last changes on https://github.com/WordPress/gutenberg/pull/66531

File: packages/fields/tsconfig.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
 		{ "path": "../private-apis" },
 		{ "path": "../router" },
 		{ "path": "../url" },
+		{ "path": "../block-editor" },
```
Comment (gigitux) [inline packages/fields/tsconfig.json:29]: We need to fix this!

File: packages/fields/package.json [hunk 2]
```diff
@@ -35,6 +35,7 @@
 		"@babel/runtime": "7.25.7",
 		"@wordpress/api-fetch": "*",
 		"@wordpress/blob": "*",
+		"@wordpress/block-editor": "*",
```
Comment (gigitux) [inline packages/fields/package.json:38]: We need to fix this!

**Comment by gigitux** [comment]
@louwie17, @youknowriad, @oandregal, I'm adding you as reviewers to gather some initial feedback on this PR.

Currently, there are two things to mention:

### `__experimentalBlockPatternsList` usage

This approach seems acceptable, but I welcome your opinions.

### `Uncaught TypeError: Cannot destructure property 'avatarURL' of '(0 , _wordpress_data__WEBPACK_IMPORTED_MODULE_3__.useSelect)(...)' as it is undefined`

The only solution appears to be removing the preloaded setting object and moving it to a REST endpoint, as we did for `template types` and `template part areas` (see https://github.com/WordPress/gutenberg/pull/66459)

File: packages/fields/src/fields/template/template-edit.tsx [hunk 3]
```diff
@@ -0,0 +1,245 @@
...
+}: DataFormControlProps< BasePost > ) => {
+	const { id } = field;
+	const postType = data.type;
+	const slug = data.slug;
+
+	const allowSwitchingTemplate = useAllowSwitchingTemplates( {
+		postType: data.type,
+		postId: typeof data.id === 'number' ? data.id : parseInt( data.id, 10 ),
+	} );
+	const templates = useTemplates( data.type );
```
Comment (louwie17) [inline packages/fields/src/fields/template/template-edit.tsx]:
```suggestion
	const templates = useTemplates( postType );
```
Given you have already defined it up above.

**Review by louwie17 — APPROVED** [review]
This tested well and changes look good from my end, I just left one minor suggestion. 

> __experimentalBlockPatternsList usage

I agree that this is fine given the template field also lives within the `wordpress/fields` package.

File: packages/edit-site/src/components/post-edit/index.js [hunk 4]
```diff
@@ -139,17 +147,26 @@ function PostEditForm( { postType, postId } ) {
...
-				'is-empty': ! postId,
-			} ) }
-			label={ __( 'Post Edit' ) }
-		>
-			{ postId && (
-				<PostEditForm postType={ postType } postId={ postId } />
-			) }
-			{ ! postId && <p>{ __( 'Select a page to edit' ) }</p> }
-		</Page>
+		<ExperimentalBlockEditorProvider settings={ settings }>
```
Comment (youknowriad) [inline packages/edit-site/src/components/post-edit/index.js]: Why not wrap the previews instead, for me this makes the dependency clearer.

File: packages/fields/src/fields/template/template-edit.tsx [hunk 5]
```diff
@@ -0,0 +1,245 @@
...
+	Modal,
+} from '@wordpress/components';
+import { useAsyncList } from '@wordpress/compose';
+import { useSelect } from '@wordpress/data';
+import { decodeEntities } from '@wordpress/html-entities';
+import { __ } from '@wordpress/i18n';
+import { getItemTitle } from '../../actions/utils';
+import type { BasePost } from '../../types';
+
+function useTemplates( postType: string ) {
```
Comment (youknowriad) [inline packages/fields/src/fields/template/template-edit.tsx]: It's not clear to me why this hook is here and not just inlined when we need the templates. It feels like it's only used once.

File: packages/fields/src/fields/template/template-edit.tsx [hunk 6]
```diff
@@ -0,0 +1,245 @@
...
+				}
+			);
+			const isPostsPage = +postId === siteSettings?.page_for_posts;
+			// If current page is set front page or posts page, we also need
+			// to check if the current theme has a template for it. If not
+			const isFrontPage =
+				postType === 'page' &&
+				+postId === siteSettings?.page_on_front &&
+				templates?.some( ( { slug } ) => slug === 'front-page' );
+			return ! isPostsPage && ! isFrontPage;
```
Comment (youknowriad) [inline packages/fields/src/fields/template/template-edit.tsx]: I think now there's a  `getHomePage` and `getPostsPageId` selectors that exits in core-data to simplify things

File: packages/fields/src/fields/template/template-edit.tsx [hunk 7]
```diff
@@ -0,0 +1,245 @@
...
+				{
+					per_page: -1,
+					post_type: postType,
+				}
+			) ?? [],
+		[ postType ]
+	);
+}
+
+export function useAllowSwitchingTemplates( {
```
Comment (youknowriad) [inline packages/fields/src/fields/template/template-edit.tsx]: Same here, is this hook really needed?

File: packages/edit-site/src/components/post-edit/index.js (hunk 4 above)
Comment (gigitux) [reply packages/edit-site/src/components/post-edit/index.js]: It makes sense. Improved with 704ebe593e604c6670ee74d20badb27439360bab

File: packages/fields/src/fields/template/template-edit.tsx (hunk 5 above)
Comment (gigitux) [reply packages/fields/src/fields/template/template-edit.tsx]: Fixed with 7cf284378c54a2fdbf0c43b9a68ff085bc0b7b21

File: packages/fields/src/fields/template/template-edit.tsx (hunk 6 above)
Comment (gigitux) [reply packages/fields/src/fields/template/template-edit.tsx]: Fixed with 7cf284378c54a2fdbf0c43b9a68ff085bc0b7b21

File: packages/fields/src/fields/template/template-edit.tsx (hunk 7 above)
Comment (gigitux) [reply packages/fields/src/fields/template/template-edit.tsx]: Fixed with 7cf284378c54a2fdbf0c43b9a68ff085bc0b7b21

**Comment by gigitux** [comment]
Thanks for the review! 
I've updated the PR. It seems that the `Cannot destructure property 'avatarURL'` issue is no longer reproducible 🤔

File: packages/edit-site/src/components/post-edit/index.js (hunk 4 above)
Comment (youknowriad) [reply packages/edit-site/src/components/post-edit/index.js]: I think you misunderstood my suggestion :) 

I was asking to wrap the field "edit" component (or render) with `ExperimentalBlockEditorProvider` instead of wrapping the whole form. None of the other fields need this provider.

File: packages/fields/src/fields/template/template-edit.tsx (hunk 7 above)
Comment (youknowriad) [reply packages/fields/src/fields/template/template-edit.tsx]: Did you actually remove this hook because I still see it in the diff :)

File: packages/fields/src/fields/template/template-edit.tsx [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+			const isPostsPage = getPostsPageId() === +postId;
+			const isFrontPage =
+				postType === 'page' && getHomePage()?.postId === +postId;
+
+			return ! isPostsPage && ! isFrontPage;
+		},
+		[ postId, postType ]
+	);
+
+	const { availableTemplates, templates } = useSelect(
```
Comment (youknowriad) [inline packages/fields/src/fields/template/template-edit.tsx:42]: Why not merge both useSelect calls?

File: packages/fields/src/fields/template/template-edit.tsx (hunk 7 above)
Comment (gigitux) [reply packages/fields/src/fields/template/template-edit.tsx]: I forgot to remove it - sorry 😅. Fixed with f891475bd3f69bcb66b1fdadff1e7173fdf5d424.

File: packages/fields/src/fields/template/template-edit.tsx (hunk 8 above)
Comment (gigitux) [reply packages/fields/src/fields/template/template-edit.tsx:42]: Fixed with f891475bd3f69bcb66b1fdadff1e7173fdf5d424

File: packages/edit-site/src/components/post-edit/index.js (hunk 4 above)
Comment (gigitux) [reply packages/edit-site/src/components/post-edit/index.js]: Sorry! Gotcha!

I fixed with cbb8bda871392548d2595f40568cc95eaf6854e4. Another good way to fix it, it is updating the `usePostFields`: https://github.com/WordPress/gutenberg/blob/f891475bd3f69bcb66b1fdadff1e7173fdf5d424/packages/editor/src/components/post-fields/index.ts#L26-L73.


The issue is that `usePatternSettings` isn't defined in the `@wordpress/editor` package. I'm happy to proceed with a refactor if you think that it makes sense.

File: packages/edit-site/src/components/post-edit/index.js [hunk 9]
```diff
@@ -123,14 +131,39 @@ function PostEditForm( { postType, postId } ) {
...
+	 * The template field depends on the block editor settings.
+	 * This is a workaround to ensure that the block editor settings are available.
+	 */
+	const fieldsWithDependency = useMemo( () => {
+		return fields.map( ( field ) => {
+			if ( field.id === 'template' ) {
+				return {
+					...field,
+					Edit: ( data ) => (
+						<ExperimentalBlockEditorProvider settings={ settings }>
```
Comment (youknowriad) [inline packages/edit-site/src/components/post-edit/index.js:150]: Why is this here and not directly in the template field edit?

Comment (youknowriad) [reply packages/edit-site/src/components/post-edit/index.js:150]: Not asking to change it, just curious why

Comment (gigitux) [reply packages/edit-site/src/components/post-edit/index.js:150]: I need to pass `settings` to this provider and `usePatternSettings` depends on the edit site store:

https://github.com/WordPress/gutenberg/blob/f891475bd3f69bcb66b1fdadff1e7173fdf5d424/packages/edit-site/src/components/page-patterns/use-pattern-settings.js#L15-L20

Comment (youknowriad) [reply packages/edit-site/src/components/post-edit/index.js:150]: Ok, here are my thoughts on this. Not necessarily something to be done here though

We need to check what settings exactly are needed for these previews. It seems mostly about the list of available patterns (maybe there's other stuff but I'm not sure). IMO, these could be provided directly from `core-data` package selectors... without the "settings" object. This is a cleanup that needs to be done at some point. I mention some of this here https://make.wordpress.org/core/2024/09/12/gutenberg-development-practices-and-common-pitfalls/ (It's also kind of a similar change to what we did for defaultTemplates and template parts)

I'm ok shipping this as is now but I feel like if someone uses the "template" field, it should just work without any hidden requirement.

Comment (gigitux) [reply packages/edit-site/src/components/post-edit/index.js:150]: Thanks! I agree with you! I'm happy to:
- open an issue.
- add the issue link on this comment.
- dedicate some time in the upcoming weeks to fix this.

Comment (youknowriad) [reply packages/edit-site/src/components/post-edit/index.js:150]: Awesome ❤️

**Review by youknowriad — APPROVED** [review]

File: packages/fields/README.md [hunk 10]
```diff
@@ -98,6 +98,10 @@ Undocumented declaration.
 
 Status field for BasePost.
 
+### templateField
+
+Undocumented declaration.
```
Comment (oandregal) [inline packages/fields/README.md:103]: We should aim to document every export. It's quite discouraging to look at docs that say "Undocumented declaration".

File: packages/fields/src/fields/template/template-edit.tsx [hunk 11]
```diff
@@ -0,0 +1,210 @@
+/**
```
Comment (oandregal) [inline packages/fields/src/fields/template/template-edit.tsx:1]: For file names: we could simplify to `edit` or `view` if they are within a field folder.

File: packages/fields/src/fields/template/index.ts [hunk 12]
```diff
@@ -0,0 +1,22 @@
...
+ */
+import { __ } from '@wordpress/i18n';
+import type { BasePost } from '../../types';
+import { TemplateEdit } from './template-edit';
+
+const templateField: Field< BasePost > = {
+	id: 'template',
+	type: 'text',
+	label: __( 'Template' ),
+	getValue: ( { item } ) => item.template,
```
Comment (oandregal) [inline packages/fields/src/fields/template/index.ts:17]: We don't need to provide a `getValue` function here given id is `template`.

**Comment by Mamaduka** [comment]
I noticed that changes from this PR were triggering `useSelect` warnings. Here's the fix #69344.
